The great tlv wire migration of 2021

halseth opened the pull request:
This is a continuation/slight rewrite of `multi: the great tlv wire migration` #4747.

There are a few key differences:
- instead of making the `lnwire` package version aware in order to use it in migrations for the old format, we copy the legacy lnwire code into the migration package. This allow us to change `lnwire` without worrying about breaking migrations.
- ~~Instead of having `UpfrontShutdownScript` be its own field on the `[Open|Acccept]Channel` messages, we embed it as a proper TLV field withing the `ExtraData`, and fetch it from there when needed.~~ We are back to having a separate field.

Replaces #4747

halseth commented on channeldb/migration21/migration.go:
```diff
@@ -184,18 +200,23 @@ func MigrateDatabaseWireMessages(tx kvdb.RwTx) error {
+	closedChanBucket := tx.ReadWriteBucket(closedChannelBucket)
+
+	// Exit early if bucket is not found.
+	if closedChannelBucket == nil {
```
Change here is that we exit early instead of returning error if bucket is not found.

Roasbeef commented on channeldb/channel.go:
```diff
@@ -1940,12 +1940,12 @@ func deserializeLogUpdates(r io.Reader) ([]LogUpdate, error) {
 	}
 
-	if err := diff.CommitSig.Encode(w, 0); err != nil {
+	if err := WriteElements(w, diff.CommitSig); err != nil {
```
For the other reviewer, this is where the encode/decode change actually happens: we'll now use `WriteElements` which'll use the proper `lnwire` message encoding (including the type and length prefix) instead of `CommitSig.Encode` which writes out the raw bytes in-place.

Roasbeef commented on channeldb/codec.go:
```diff
@@ -178,7 +179,17 @@ func WriteElement(w io.Writer, element interface{}) error {
 
 	case lnwire.Message:
-		if _, err := lnwire.WriteMessage(w, e, 0); err != nil {
+		var msgBuf bytes.Buffer
```
Also here re the new length prefix before all elements.

Roasbeef reviewed (commented):
Many thanks for taking over this PR! 

The end diff _looks_ large, but it's mostly code movement, with the core logic migration logic residing in a single commit. In the end, copying the `lnwire` package is def the way to go, to ensure we have proper isolation. 

As the PR changed slightly, I'll start to test this out on on eof my nodes, much easier now that I don't need to manually creates a new breanch that combines the `lnwire` TLV changes along with the migration itself.

Roasbeef commented:
The dry run migration went well, but upon boot I get:
```
2021-02-18 21:05:58.628 [ERR] LTND: unable to create server: unable to parse message of unknown type: <unknown>
```

I dug in a bit and turns out the migration isn't using the _proper_ encoding when writing all the data back again. It uses what's in `enclosed_codec.go`, which is the _old_ encode/decode logic. Namely, it's missing the new logic for encoding/decoding `lnwire.Message` w/ the proper length prefix. 

I have a commit that fixes this, but before pushing it up, I want to se why this _wasn't_ picked up by the existing tests.

Roasbeef commented:
Here's the commit: https://github.com/lightningnetwork/lnd/commit/239f19e4bd3b6a7db5c460b3f679c910d06fc974

Also the reason why the tests passed was that since the migration logic was using the same encode/decode as the legacy logic (before the length prefix), it wasn't actually doing anything. One way to detect this in the future is to assert in the tests that the bytes are actually _different_.

Roasbeef commented:
Also I realized now where I left off last w/ it all: migrating the forwarding packages as well. As is, the migration will try to read out the forwarding packages in `loadHltcs`, using the `LogUpdate.Decode` method, which uses `ReadElements`. This'll fail as we'll read two more extra bytes than is actually encoded, as we're trying to read the length prefix, but one isn't present there yet.

halseth commented:
> Here's the commit: [239f19e](https://github.com/lightningnetwork/lnd/commit/239f19e4bd3b6a7db5c460b3f679c910d06fc974)
> 
> Also the reason why the tests passed was that since the migration logic was using the same encode/decode as the legacy logic (before the length prefix), it wasn't actually doing anything. One way to detect this in the future is to assert in the tests that the bytes are actually _different_.

Wow, can't believe I missed that! This migration wasn't really doing much that's for sure 🤦  



> Also I realized now where I left off last w/ it all: migrating the forwarding packages as well. As is, the migration will try to read out the forwarding packages in `loadHltcs`, using the `LogUpdate.Decode` method, which uses `ReadElements`. This'll fail as we'll read two more extra bytes than is actually encoded, as we're trying to read the length prefix, but one isn't present there yet.

Will to a sweep and make sure there aren't more places where we missed this!

halseth commented:
I updated the PR to address the concerns brought up:

I split `legacy` and `current` serialization code used by the migration into two different pacakges, such that we should avoid calling into the wrong method unintended.

I went through the code base looking for all places where a wire message is written to the DB, found two more not already covered. One of them being the forwarding packages @Roasbeef mentioned, the other one the remote unisigned log updates that was recently added: https://github.com/lightningnetwork/lnd/pull/4923/commits/ea91b35fd1c0addb3b01384d60f31ce524eb9623

In addition I looked for places where the wire messages gets written to the DB directly, outside the `codec`. Here I also found to cases, but since they both are written under their own key, we should be fine (but I wouldn't mind also migrating them just to be consistent): 
https://github.com/lightningnetwork/lnd/blob/203a669578e338667b346d7c7125c8e27b04f927/discovery/message_store.go#L123
https://github.com/lightningnetwork/lnd/blob/203a669578e338667b346d7c7125c8e27b04f927/channeldb/waitingproof.go#L235

Roasbeef commented:
> In addition I looked for places where the wire messages gets written to the DB directly, outside the codec. Here I also found to cases, 

Yeah I identified those in the past as well, but left them as they're written under an isolated key.

Roasbeef commented:
Started up with the latest diff, and so far so good 👌

Roasbeef commented:
Accidentally closed...

Roasbeef commented on channeldb/channel.go:
```diff
@@ -1791,14 +1791,19 @@ type LogUpdate struct {
-// Encode writes a log update to the provided io.Writer.
-func (l *LogUpdate) Encode(w io.Writer) error {
+// serializeLogUpdate writes a log update to the provided io.Writer.
+func serializeLogUpdate(w io.Writer, l *LogUpdate) error {
```
👍

Roasbeef commented on funding/manager.go:
```diff
@@ -1388,10 +1388,29 @@ func (f *Manager) handleFundingOpen(peer lnpeer.Peer,
 	reservation.SetOurUpfrontShutdown(shutdown)
 
+	// Pack it into the ExtraData of the message as a TLV record.
+	var extraData lnwire.ExtraOpaqueData
```
Hmm I see how this jumps forward to make the fields "TLV" native, but I think it may be better to hide more or this information from the caller, the way it was in the prior diff. So just setting the actual _fields_ on the struct, and letting the `Encode/Decode` methods do the rest.

As now we've fored the funding manager to keep up with all the future TLV types vs hiding the details behind `Encode/Decode` and using `nil`ness to indicate if a field was set or not.

Roasbeef commented on lnwire/accept_channel.go:
```diff
@@ -145,18 +144,28 @@ func (a *AcceptChannel) Decode(r io.Reader, pver uint32) error {
+func (a *AcceptChannel) UpfrontShutdownScript() (DeliveryAddress, error) {
+	var addr DeliveryAddress
+
+	tlvs, err := a.ExtraData.ExtractRecords(addr.NewRecord())
```
Related to the comment above, we could have the callers use methods to "attach" these fields, which exposes less details to them, and is an alternative to setting `nil` values. I still loosely favor letting the callers set the fields themselves, and hiding everything behind `Encode/Decode` re if it's actually a TLV or not. As we evolve the messages, I imagine we'll have more and more "TLV native" fields, which can start to clutter the call sites if they need to always manually pack/encode them.

Roasbeef commented on lnwire/accept_channel.go:
```diff
@@ -145,18 +144,28 @@ func (a *AcceptChannel) Decode(r io.Reader, pver uint32) error {
+func (a *AcceptChannel) UpfrontShutdownScript() (DeliveryAddress, error) {
+	var addr DeliveryAddress
+
+	tlvs, err := a.ExtraData.ExtractRecords(addr.NewRecord())
```
Related to the comment above, we could have the callers use methods to "attach" these fields, which exposes less details to them, and is an alternative to setting `nil` values. I still loosely favor letting the callers set the fields themselves, and hiding everything behind `Encode/Decode` re if it's actually a TLV or not. As we evolve the messages, I imagine we'll have more and more "TLV native" fields, which can start to clutter the call sites if they need to always manually pack/encode them.

halseth replied:
If the caller alters the exposed field without modifying the blob, the extra data blob gets out of sync with the field. A possible solution is to make the `extraData` field private, and only create it at time of encoding, packing the set fields in the process. The problem with this however, is that fields unknown to us still must be stored in the blob (imagine reading a message with a blob having unknown fields, and then attempting to add the `UpfrontShutdown` field to it. At this point we would have to add the data to the blob when encoding it, which is currently not that easy (or something we want?). We also don't know for sure whether the field is already present in the blob.

So I think there are a few avenues we can go down:

* (In this PR currently) Making the caller TLV aware such that the whole extra data blob must be set by the caller.
* Only exposing the `UpfrontShutdownField` and not the blob, and documenting that altering this field  _will invalidate_ any data already in the blob. At encoding we discard what's already in the blob and create a new one.
* Adding `Setters/Getters` for the field. The getter will simply try to fetch it from the blob, the `Setter` will add it to the blob (if the blob is non-empty with unknown fields we must either return and error at this point, or extend our TLV code to allow adding fields to an existing stream).
* Create a constructor that takes the optional fields. This way we can make sure the fields cannot be altered, and the blob cannot change after creation.

All of these are just ways to avoid the field getting out of sync with the blob of course. If we make sure to _never alter_ the field after creation, then it doesn't matter that much, but would be nice to have it somewhat user-mistake proof .

Thoughts? cc @wpaulino

Roasbeef replied:
> Only exposing the UpfrontShutdownField and not the blob, and documenting that altering this field will invalidate any data already in the blob. At encoding we discard what's already in the blob and create a new one.

I favor this option personally. I don't think we should idle long on this though, either direction can easily be reversed. I'm mostly imagining the future messages related to dynamic commitments, which will likely be _entirely_ TLV based. If we keep the current API through those messages then we'll end up with a bunch of boiler plate just to create the message.

Roasbeef reviewed (commented):
It lives!!!

Really close now, final comment is re if we should expose the TLV details (as is now), or attempt to hide more of them from the user (as it was before).

Crypt-iQ commented:
One small comment: `WriteMessage` in lnwire/message.go should check against `MaxMsgBody` instead of `MaxMessagePayload`

wpaulino reviewed (commented):
Looks pretty good! Confirmed there aren't any other unaddressed cases within the codebase and also had a successful migration on my node.

I agree with @Roasbeef regarding not exposing the TLV details to callers, it seems much cleaner. Imagine a scenario where a message goes through multiple subsystems and each subsystem attaches its own TLV field. With the current approach, both subsystems would need to be aware of all TLVs when not required, potentially causing import cycles.

halseth commented:
> Imagine a scenario where a message goes through multiple subsystems and each subsystem attaches its own TLV field.

This would be pretty cool functionality. Looks like our TLV code doesn't currently supports adding fields to an existing stream though, maybe that is something we want? cc @cfromknecht

Roasbeef commented:
> Looks like our TLV code doesn't currently supports adding fields to an existing stream though, maybe that is something we want? 

I don't think that's required. The encoding would happen only at the very end, so callers can add new TLV (which are just fields), then they're all assembled into a single stream and encoded at the end.

halseth commented:
Update the PR to keep the `shutdown script` as a separate field again. 

To make this work we read out all TLV data, then snip out the script bytes (that **MUST** be there) and set the script field on the struct: https://github.com/lightningnetwork/lnd/pull/4923/commits/7ec3067a4196ffc98c4cb26346b835ff84509e1a#diff-682634355d285b253b49554f4ea4a4a15ce3a07d98b55bb978944bb5e0e62ab7R235-R245

When writing the message, we again concatenate the script field with the rest of the data.

This enables us to change the shutdown script on the struct without invalidating the extra data in the message.

The only case that wont be entirely correct doing it this way, is if we read a legacy message (that have no upfront shutdown script set at all), then write it out again. In this case the encoding would change since we add the zero-length field in that case. But I figured this was okay, since we were doing this already before this PR.

cfromknecht commented on discovery/syncer.go:
```diff
@@ -753,7 +753,9 @@ func (g *GossipSyncer) synchronizeChanIDs() (bool, error) {
 	reply *lnwire.ReplyChannelRange) bool {
 
-	return reply.QueryChannelRange == *query
+	return (reply.ChainHash == query.ChainHash &&
```
unnecessary parens?

cfromknecht commented on lnwire/update_fulfill_htlc.go:
```diff
@@ -75,8 +82,7 @@ func (c *UpdateFulfillHTLC) MsgType() MessageType {
 func (c *UpdateFulfillHTLC) MaxPayloadLength(uint32) uint32 {
-	// 32 + 8 + 32
-	return 72
+	return MaxMsgBody
```
this entire method can be removed from the interface, since every message now has a constant max length

cfromknecht commented on lnwire/accept_channel.go:
```diff
@@ -150,15 +171,82 @@ func (a *AcceptChannel) Decode(r io.Reader, pver uint32) error {
 		return err
 	}
+
+	a.UpfrontShutdownScript, a.ExtraData, err = parseShutdownScript(
```
i think it makes sense to use `ExtraOpaqueData` deserialization for those cases where we don't have any defined TLV fields, but in cases where we have known fields I had imagined using a custom `tlv.Stream` that would  container a pointer to the fields that need to be [de]serialized. In the long run, i don't think it will be scalable to have a `parseXXX` field for everything we want to end up extracting

wpaulino replied:
What about something like:

```
// Assumes all records being parsed start at the beginning of the TLV stream and follow one another.
// If this isn't the case, we'd need the custom `tlv.Stream` that contains pointers to each record.
func (e *ExtraOpaqueData) ParseKnownRecords(producers ...tlv.RecordProducer) error {
	for _, p := range producers {
		r := p.Record()
		records, err := e.ExtractRecords(r)
		if err != nil {
			return err
		}
		rawRecordVal, ok := records[r.Type()]
		if !ok || rawRecordVal != nil {
			return fmt.Errorf("record with type %v not parsed",
				r.Type())
		}

		// Remove the bytes corresponding to the record we just parsed
		// from our TLV stream.
		parsedRecord := p.Record()
		valSize := parsedRecord.Size()
		lenSize := tlv.VarIntSize(valSize)
		typeSize := tlv.VarIntSize(uint64(parsedRecord.Type()))
		*e = []byte(*e)[typeSize+lenSize+valSize:]
	}

	return nil
}
```

In the case of `lnwire.OpenChannel`, it'd be used as:

```
	var tlvRecords ExtraOpaqueData
	if err := ReadElements(r, &tlvRecords); err != nil {
		return err
	}
	err = tlvRecords.ParseKnownRecords(&o.UpfrontShutdownScript)
	if err != nil {
		return err
	}
	o.ExtraData = tlvRecords
```

halseth replied:
I agree that this approach won't be scalable in the generic case, but the shutdown script is really a _special case_, since it must always be there. 

I think we should defer making this more generic until we have actual new fields we want to add, we'll know more about the requirements then.

Roasbeef replied:
Agreed, we can move ahead with the bulk of this PR given once we are properly writing/parsing all the data, we can make any future in-memory TLV representation changes at a later point.

cfromknecht reviewed (commented):
@halseth nice work!! Even though it is a lot of code we need to keep around, I find this diff easy to follow 👍

> This would be pretty cool functionality. Looks like our TLV code doesn't currently supports adding fields to an existing stream though, maybe that is something we want? 

That would definitely be a nice addition, shouldn't be too hard to add that.

halseth commented on lnwire/accept_channel.go:
```diff
@@ -150,15 +171,82 @@ func (a *AcceptChannel) Decode(r io.Reader, pver uint32) error {
+	}
+
+	// Not among TLV records, this means the data was invalid.
+	if _, ok := tlvs[DeliveryAddrType]; !ok {
```
Hm, looks like that doc is not entirely correct? @cfromknecht

Roasbeef replied:
This is the output of `DecodeWithParsedTypes`, which'll tally all the types parsed as it goes along: https://github.com/lightningnetwork/lnd/blob/master/tlv/stream.go#L234-L238

If the type is known, it'll be a nice slice (as the TLV record will be populated). If the type is unknown, but odd, then we'll retain the bytes in the map.

halseth replied:
I see. I don't think it should matter then, since we are only checking a known type, so we'll know that it is nil always.

Roasbeef commented on lnwire/accept_channel.go:
```diff
@@ -150,15 +171,82 @@ func (a *AcceptChannel) Decode(r io.Reader, pver uint32) error {
 
+// packShutdownScript takes an upfront shutdown script and an opaque data blob
+// and concatenates them.
+func packShutdownScript(addr DeliveryAddress, extraData ExtraOpaqueData) (
```
Don't think this is non-blocking for this PR, once we start to add more TLV fields to messages, and also messages that are purely TLV, we'll have a better idea w.r.t ergonomics of the main interface.

Roasbeef commented on lnwire/accept_channel.go:
```diff
@@ -104,6 +116,15 @@ var _ Message = (*AcceptChannel)(nil)
 func (a *AcceptChannel) Encode(w io.Writer, pver uint32) error {
+	// Since the upfront script is encoded as a TLV record, concatenate it
+	// with the ExtraData, and write them as one.
+	tlvRecords, err := packShutdownScript(
```
👍

Roasbeef reviewed (approved):
LGTM 🍙

Updated another older testnet node, w/ this migration, and it's been running fine w/ no problems for the past few days. This unblocks a ton of other projects, so really excited to finally get this in!
